auto_round/utils.py

```diff
@@ -1136,6 +1141,130 @@ def get_device_and_parallelism(device):
+            state_dict.pop(name, None)
+    gc.collect()
+    for name, module in model.named_modules():
+        if hasattr(module, "weight_scale"):
+            count+=1
+            state_dict[f"{name}.weight_scale"] = module.weight_scale
+            state_dict[f"{name}.weight"] = state_dict[f"{name}.weight"].to(torch.int8)
+            gc.collect()
+    print(f"quantized_count: {count}")
+    
+    # handle specific large experts
+    new_state_dict = {}
+    from tqdm import tqdm
+    state_list = list(state_dict.keys())
+    for name in tqdm(state_list):
+        if name.endswith("_fake.weight"):
+            weight = state_dict[name]
+            if weight.dim() != 3:
+                continue  # skip any unexpected format
+            for id in range(int(weight.size(0))):
+                scale_name = f"{name}_scale"
+                weight_name_expert = name.replace("_fake", "")
+                weight_name_expert = weight_name_expert.replace("experts.", "experts."+str(id)+".")
+                weight_expert = weight[id].transpose(0,1).contiguous()
+                scale_expert = state_dict[scale_name][id]
+                if "gate_up_proj" in name:
+                    weight_expert_0, weight_expert_1 = weight_expert.chunk(2,dim=0)
+                    weight_expert_0 = weight_expert_0.contiguous()
+                    weight_expert_1 = weight_expert_1.contiguous()
+                    scale_0, scale_1 = scale_expert.chunk(2)
+                    scale_0 = scale_0.contiguous()
+                    scale_1 = scale_1.contiguous()
+                    weight_name_expert_0 = weight_name_expert.replace("gate_up_proj", "gate_proj")
+                    weight_name_expert_1 = weight_name_expert.replace("gate_up_proj", "up_proj")
+                    new_scale_name_0 = f"{weight_name_expert_0}_scale"
+                    new_scale_name_1 = f"{weight_name_expert_1}_scale"
+                    new_state_dict[weight_name_expert_0] = weight_expert_0
+                    new_state_dict[new_scale_name_0] = scale_0
+                    new_state_dict[weight_name_expert_1] = weight_expert_1
+                    new_state_dict[new_scale_name_1] = scale_1
+                else:
+                    new_scale_name = f"{weight_name_expert}_scale"
+                    new_state_dict[weight_name_expert] = weight_expert
+                    new_state_dict[new_scale_name] = scale_expert
+            state_dict.pop(name, None)
+            state_dict.pop(scale_name, None)
+            gc.collect()
+        elif ".experts." not in name:
+            new_state_dict[name] = state_dict[name]
+        else:
+            continue
+    return new_state_dict
+
+
+def pack_to_int8(model, output_dir):
+    import json
+    from safetensors.torch import save_file
+    with torch.no_grad():
+        state_dict = translate_2_sglang_int8(model)
+    max_shard_size = 40 * 1024**3  # 40GB
+    shards = {}
+    current_shard = {}
```

Comment:
this part should be refined and supported in main branch.
Better follow the origin code style if possible